In [3]:
import pandas as pd 
df = pd.read_csv(r"C:\Users\ASUS\Downloads\Advanced_pandas_sales_dataset.csv")
df.head()

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30


In [4]:
#Q1 Dataset Exploration Perform the following exploratory steps: 
#Display the first 10 rows 
df.head(10)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [5]:
#Display the dataset structure using info() 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB


In [7]:
#Find the number of rows and columns
df.shape

(230, 11)

In [9]:
#Identify the data types of each column
df.dtypes

Order_ID        int64
Order_Date     object
Customer       object
City           object
Region         object
Product        object
Category       object
Sales           int64
Quantity        int64
Discount      float64
Profit        float64
dtype: object

In [11]:
#Q2 Sales Filtering Analysis Filter the dataset to display:
#1.Orders where Sales > 3500 
df[df['Sales'] > 3500]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
10,1011,2024-06-30,Neha,Delhi,South,Printer,Office,5977,6,0.00,1578.32
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [12]:
#2.Orders where Region = "East" and Sales > 2500 
df[(df['Region'] == 'East') & (df['Sales'] > 2500)]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
13,1014,2024-03-18,Anita,Bangalore,East,Mouse,Accessories,3616,4,0.05,463.93
21,1022,2024-01-14,Nitin,Chennai,East,Monitor,Electronics,4827,6,0.05,912.34
27,1028,2024-07-21,Dev,Kolkata,East,Tablet,Electronics,4054,6,0.00,1013.83
33,1034,2024-12-15,Nitin,Ahmedabad,East,Tablet,Electronics,4378,2,0.20,-104.30
34,1035,2024-02-20,Nitin,Ahmedabad,East,Laptop,Electronics,2535,4,0.10,188.15
41,1042,2024-07-15,Rohan,Chennai,East,Laptop,Electronics,3268,8,0.15,295.15
43,1044,2024-09-20,Kavya,Hyderabad,East,Printer,Office,2618,7,0.05,454.64


#Explain what insights you observe.
1.High Sales Analysis (Sales > 3500)
Electronics Dominance: The majority of high-value orders come from the Electronics category, specifically Phones and Monitors. This suggests that high-revenue growth is driven primarily by tech hardware.

Profit Volatility: High sales do not always guarantee high profit. For instance, Order 1004 (Keyboard) shows a loss of -71.66 despite high sales, likely due to the 20% discount applied.

Customer Loyalty: Kavya and Meera appear multiple times in the high-sales list, indicating they might be "Power Users" or corporate clients who place large bulk orders.

2.Regional Analysis (East Region & Sales > 2500)
Heavy Discounting in East: Several orders in the East region (like Order 1034 and Order 1088) are resulting in negative profit. This indicates that the sales team in the East may be over-relying on high discounts (20%) to close large deals.

Product Preferences: In the East, Tablets and Laptops are frequent high-value items. This could suggest a strong market for portable computing in cities like Kolkata and Ahmedabad.

Efficiency: Orders with 0.00 to 0.05 discounts (like Order 1008 or 1028) show significantly higher profit margins (e.g., 918.49 and 1013.83), proving that the East region is highly profitable when discounts are kept below 10%.

In [13]:
#Q3 Customer Purchase Analysis Identify the top 5 customers based on total sales. Hint: Use GroupBy with aggregation.
df.groupby('Customer')['Sales'].sum().sort_values(ascending = False).head(5)

Customer
Nitin    59820
Raj      58416
Rahul    52291
Dev      48500
Isha     46818
Name: Sales, dtype: int64

In [16]:
#Q4 Regional Sales Performance Calculate the total sales generated by each region. 
#Then identify: Which region generated the highest revenue?

df.groupby('Region')['Sales'].sum().sort_values(ascending = False)
df.groupby('Region')['Sales'].sum().nlargest(1)

Region
East    232675
Name: Sales, dtype: int64

In [17]:
#Q5 Product Performance Analysis 
#Calculate the total quantity sold for each product. 
df.groupby('Product')['Quantity'].sum().sort_values(ascending=False)

Product
Monitor       156
Phone         151
Tablet        150
Laptop        140
Mouse         122
Headphones    115
Keyboard      110
Printer       108
Name: Quantity, dtype: int64

In [18]:
#Which product appears to be the most popular?
df.groupby('Product')['Quantity'].sum().idxmax()

'Monitor'

In [19]:
#Q6 Multi-Level Grouping 
#Group the dataset by: Region and Category
#Calculate the total sales for each combination. 
df.groupby(["Region","Category"])["Sales"].sum().sort_values(ascending = False)

Region  Category   
East    Electronics    160907
South   Electronics     86690
West    Electronics     84128
South   Accessories     80914
North   Electronics     70023
West    Accessories     64811
North   Accessories     62052
East    Accessories     57646
North   Office          26135
West    Office          21288
South   Office          15504
East    Office          14122
Name: Sales, dtype: int64

#What insights can you derive from this analysis?
1.Dominance of the East-Electronics Segment The most significant finding is that East-Region Electronics is the primary revenue driver, generating 160,907 in sales. This single combination outperforms the next highest segment (South-Electronics) by nearly 2x, indicating a massive market concentration in the East for tech products.

2.Category Performance Hierarchy There is a clear hierarchy in product demand across all regions:
Electronics is the top-performing category in every single region.
Accessories consistently holds the middle ground.
Office Supplies is the lowest revenue generator across all regions. This suggests the business's core strength lies in high-value tech hardware rather than general administrative supplies.

3.Regional Specialization in Accessories While the East leads in Electronics, the South and West regions show strong performance in Accessories (both exceeding 64,000). In fact, the South's Accessory sales (80,914) are nearly as high as its Electronics sales, suggesting a more balanced product mix in the Southern market compared to the East.

4.Underperforming Segments & Opportunities The Office category shows the lowest engagement, particularly in the East (14,122) and South (15,504). Given that the East has the highest overall customer traffic, there is a clear strategic opportunity to improve cross-selling tactics to convert Electronics buyers into Office Supply customers.

In [21]:
#Q7 Pivot Table Creation 
#Create a pivot table showing: 
#Rows → Region 
#Columns → Product 
#Values → Total Sales 
#Which region sells the most laptops?
pivot_table = df.pivot_table(index='Region', columns='Product', values='Sales', aggfunc='sum')


display(pivot_table)

top_laptop_region = pivot_table['Laptop'].idxmax()
print(f"The region that sells the most laptops is: {top_laptop_region}")

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274


The region that sells the most laptops is: East


In [24]:
#Q8 Profitability Analysis 
#Calculate the average profit for each product category.
avg_profit = df.groupby("Category")["Profit"].mean().sort_values(ascending = False)
print("Average Profit per Category:")
display(avg_profit)
#Which category is most profitable?
most_profitable_category = avg_profit.idxmax()
print(f"\nThe most profitable category is: {most_profitable_category}")

Average Profit per Category:


Category
Office         471.938462
Accessories    410.603086
Electronics    396.462114
Name: Profit, dtype: float64


The most profitable category is: Office


In [26]:
# Q9 Discount Impact Study Analyze the relationship between discount and profit. 
    #Tasks: 
   # 1.Calculate the average profit for each discount level 
   
discount_analysis = df.groupby('Discount')['Profit'].mean().sort_values(ascending=False)

print("Average Profit by Discount Level:")
display(discount_analysis)

# 2.Identify whether higher discounts lead to lower profit
correlation = df['Discount'].corr(df['Profit'])
print(f"\nCorrelation between Discount and Profit: {correlation:.2f}")

Average Profit by Discount Level:


Discount
0.00    751.210588
0.05    531.368039
0.10    346.460513
0.15    237.191400
0.20     90.026923
Name: Profit, dtype: float64


Correlation between Discount and Profit: -0.66


In [27]:
#Q10 End-to-End Data Wrangling Task Perform the following workflow: 
#Filter the dataset where Sales > 3000 
#Group data by Region and Product 
#Calculate total sales Reshape the result into a pivot table 

high_value_df = df[df['Sales'] > 3000]


wrangled_pivot = high_value_df.pivot_table(index='Region', 
                                           columns='Product', 
                                           values='Sales', 
                                           aggfunc='sum').fillna(0)

print("High-Value Sales Pivot Table (Sales > 3000):")
display(wrangled_pivot)

High-Value Sales Pivot Table (Sales > 3000):


Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,14469,9642,44096,16420,19000,29500,3696,29229
North,5326,15653,4540,5087,22209,24670,17185,3392
South,10026,19249,17950,14921,26928,18218,14143,19613
West,13898,33380,5922,14354,3511,15653,12648,17185


Explain the business insights obtained from this analysis.
1.Regional Performance Drivers East Region (Electronics Dominance): The East is the strongest market for high-ticket electronics, specifically Laptops (44,096) and Tablets (29,229). This region should be the primary focus for premium product launches.
West Region (Peripheral Specialist): While the West has lower Laptop sales, it leads the country in high-value Keyboards (33,380). This suggests a niche market of power users or gamers in this region.

2.Product-Specific Trends High-Value Peripherals: The Mouse category shows surprisingly high revenue in the South (26,928) and North (22,209). These aren't typically expensive items, implying these regions are likely purchasing premium gaming mice or buying in bulk "bundles" that exceed the 3,000 threshold.
Printer Market Gap: High-value Printers are underperforming in the East (3,696) compared to the North (17,185). There is a clear opportunity to expand the premium printing segment in the Eastern market.

3.Data Conclusion The analysis reveals that "High-Value" sales are not uniform. The East relies on big-ticket core hardware (Laptops/Tablets), while the North and South drive significant revenue through high-end accessories (Mice/Keyboards). This indicates that a "one-size-fits-all" marketing strategy across India would be inefficient.